app/src/test/java/org/simple/clinic/analytics/TimingCollectorTest.kt

```diff
@@ -0,0 +1,306 @@
+            .plusMinutes(1L))
+    val timeTakenForOperation3 = Duration.between(operation3StartedAt, operation3StoppedAt)
+    assertThat(timingCollector.completed.keys)
+        .isEqualTo(setOf(stage1AnalyticsName, stage3AnalyticsName))
+    assertThat(timingCollector.completed[stage3AnalyticsName])
+        .isEqualTo(timeTakenForOperation3)
+
+    clock.advanceBy(Duration.ofMinutes(30L))
+    timingCollector.stop(stage2)
+    assertThat(timingCollector.ongoing.keys).isEmpty()
+    val operation2StoppedAt = operation1StartedAt.plus(
+        Duration
+            .ofMinutes(1L)
+            .plusHours(2L)
+            .plusMinutes(1L)
+            .plusMinutes(30L))
+    val timeTakenForOperation2 = Duration.between(operation2StartedAt, operation2StoppedAt)
+    assertThat(timingCollector.completed.keys)
+        .isEqualTo(setOf(stage1AnalyticsName, stage3AnalyticsName, stage2AnalyticsName))
+    assertThat(timingCollector.completed[stage2AnalyticsName])
+        .isEqualTo(timeTakenForOperation2)
+
+    val expectedRunId = timingCollector.runId
+    val expected = listOf(
+        MockAnalyticsReporter.Event(
+            name = "TimeTaken",
+            props = mapOf(
+                "operationName" to stage1AnalyticsName,
+                "runId" to expectedRunId,
+                "timeTakenMillis" to timeTakenForOperation1.toMillis()
+            )
+        ),
+        MockAnalyticsReporter.Event(
+            name = "TimeTaken",
+            props = mapOf(
+                "operationName" to stage3AnalyticsName,
+                "runId" to expectedRunId,
+                "timeTakenMillis" to timeTakenForOperation3.toMillis()
+            )
+        ),
+        MockAnalyticsReporter.Event(
+            name = "TimeTaken",
+            props = mapOf(
+                "operationName" to stage2AnalyticsName,
+                "runId" to expectedRunId,
+                "timeTakenMillis" to timeTakenForOperation2.toMillis()
+            )
+        )
+    )
+
+    assertThat(reporter.receivedEvents).isEqualTo(expected)
+  }
+
+  @Test
+  fun `stopping a stage that does not exist must not fail`() {
+    val reporter = MockAnalyticsReporter()
+    Analytics.addReporter(reporter)
+
+    assertThat(timingCollector.ongoing).isEmpty()
+    timingCollector.stop("Operation")
+    assertThat(timingCollector.completed).isEmpty()
+
+    assertThat(reporter.receivedEvents).isEmpty()
+  }
+
+  @Test
+  fun `starting a stage that already exists must override it`() {
+    val now = Instant.now(clock)
+
+    val stageName = "Sub Operation"
+    timingCollector.start(stageName)
+    val advanceClockBy = Duration.ofHours(1L)
+    clock.advanceBy(advanceClockBy)
+    timingCollector.start(stageName)
+
+    assertThat(timingCollector.completed).isEmpty()
+    assertThat(timingCollector.ongoing["$name:$stageName"])
+        .isEqualTo(now.plus(advanceClockBy))
+  }
+
+  @Test
+  fun `starting and stopping a stage that has already completed must override it`() {
```

Comment:
AFAIK, both the duration for that stage will be reported to Heap and not overridden. Is that right?
Hence the word `override` is misleading here.

Reply:
Can we also add an assertion to check what the `totalTimeTakenForOperation` would be when this operation is finished?

Reply:
> AFAIK, both the duration for that stage will be reported to Heap and not overridden. Is that right?
> Hence the word `override` is misleading here.

Yes, I'll rework this a bit.

Reply:
I have removed the `.finish()` operation because it was getting hard to reason about what it's behaviour should be. I examined the reason why I added it, and I realised it can be replaced with the usage of another `OperationTimingTracker` and this makes things a lot simpler.